### PR TITLE
feat: parallelizing VID polynomial evaluation

### PR DESCRIPTION
**Pull request by mrain**
## Description

closes: #494

On my local laptop (M1 pro), with payload size `1 << 25`, `payload_chunk_size = 256` and `num_storage_nodes = 512`.
Running with `RAYON_NUM_THREADS=8`.

Previous result for polynomial evaluation:
```bash
··Start:   encode payload bytes into polynomials
··End:     encode payload bytes into polynomials ...................................899.311ms
··Start:   compute all storage node evals for 4229 polynomials with 256 coefficients
··End:     compute all storage node evals for 4229 polynomials with 256 coefficients 1.026s
```

Now we have:
```bash
··Start:   encode payload bytes into polynomials
··End:     encode payload bytes into polynomials ...................................168.235ms
··Start:   compute all storage node evals for 4229 polynomials with 256 coefficients
··End:     compute all storage node evals for 4229 polynomials with 256 coefficients 80.797ms
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Review by ggutoski — COMMENTED** [review]
looks great! Just a nit about trait bounds

File: primitives/src/vid/advz.rs [hunk 1]
```diff
@@ -235,7 +241,7 @@ where
 impl<E, H> VidScheme for Advz<E, H>
 where
     E: Pairing,
-    H: HasherDigest,
+    H: HasherDigest + Send + Sync,
```
[the remark this answers is not in the document]
Comment (mrain) [reply primitives/src/vid/advz.rs:244]: Sure, should we also require `H: HashDigest` for `Advz` struct ?

[the remark this answers is not in the document]
Comment (ggutoski) [reply primitives/src/vid/advz.rs:244]: In general we should require the minimum possible trait bounds for any given `impl` block. So no, don't require `H: HashDigest` for `Advz`.

File: primitives/src/vid/advz.rs [hunk 2]
```diff
@@ -58,6 +62,7 @@ pub mod precomputable;
 pub struct Advz<E, H>
 where
     E: Pairing,
+    H: Send + Sync,
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply primitives/src/vid/advz.rs:65]: The `Advz` struct has no fields that use `H` (other than `PhantomData`). We should be able to derive `Send`, `Sync` for `Advz` struct via some trick with `derivative` and `bound=""` or whatever, similar to how one derives serde traits for structs with generic type params.

File: primitives/src/vid/advz.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrain) [reply primitives/src/vid/advz.rs]: Oh I mistakenly thought we need `Send + Sync` here. Check [256156d](https://github.com/EspressoSystems/jellyfish/pull/496/commits/256156de9050e7e4f1d54b5b9625a05e8a76ef6c)

File: primitives/src/vid/advz.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrain) [reply primitives/src/vid/advz.rs]: Removed in [256156d](https://github.com/EspressoSystems/jellyfish/pull/496/commits/256156de9050e7e4f1d54b5b9625a05e8a76ef6c)

File: primitives/src/vid/advz/payload_prover.rs [hunk 3]
```diff
@@ -67,7 +67,7 @@ pub struct LargeRangeProof<F> {
 impl<E, H> PayloadProver<SmallRangeProof<KzgProof<E>>> for Advz<E, H>
 where
     E: Pairing,
-    H: HasherDigest,
+    H: HasherDigest + Send + Sync,
```
Comment (ggutoski) [inline primitives/src/vid/advz/payload_prover.rs:70]: you already added `Send`, `Sync` to `HasherDigest` right? So these new trait bounds should be unnecessary.

File: primitives/src/vid/advz/precomputable.rs [hunk 4]
```diff
@@ -29,7 +29,7 @@ use super::KzgCommit;
 impl<E, H> Precomputable for Advz<E, H>
 where
     E: Pairing,
-    H: HasherDigest,
+    H: HasherDigest + Send + Sync,
```
Comment (ggutoski) [inline primitives/src/vid/advz/precomputable.rs:32]: same here

**Comment by mrain** [comment]
Further improves `bytes_to_polys`, right now
```
··Start:   encode payload bytes into polynomials
··End:     encode payload bytes into polynomials ...................................44.674ms
··Start:   compute all storage node evals for 4229 polynomials with 256 coefficients
··End:     compute all storage node evals for 4229 polynomials with 256 coefficients 82.175ms
```

File: primitives/src/vid/advz/payload_prover.rs (hunk 3 above)
Comment (mrain) [reply primitives/src/vid/advz/payload_prover.rs]: removed [607ea41](https://github.com/EspressoSystems/jellyfish/pull/496/commits/607ea411f81d2719f5d20adbe6553797c83a1a44)

File: primitives/src/vid/advz/precomputable.rs (hunk 4 above)
Comment (mrain) [reply primitives/src/vid/advz/precomputable.rs]: removed [607ea41](https://github.com/EspressoSystems/jellyfish/pull/496/commits/607ea411f81d2719f5d20adbe6553797c83a1a44)

**Review by ggutoski — APPROVED** [review]
LFG! 🚀

**Pull request merged**